### PR TITLE
HeaderMap::addCopy

### PR DESCRIPTION
**Pull request by kflynn**
Implement `HeaderMap::addCopy`, which adds a header by copying both the key and the value.

### CODE REVIEW

**Comment by dnoe** [comment]
I'll review this. To fix the format I always use the CI docker container:

`./ci/run_envoy_docker.sh './ci/do_ci.sh fix_format'`

File: source/common/http/header_map_impl.h [hunk 1]
```diff
@@ -54,9 +54,17 @@ class HeaderMapImpl : public HeaderMap {
   bool operator==(const HeaderMapImpl& rhs) const;
 
   // Http::HeaderMap
+  // Add without copying either key or value.€
```
Comment (kflynn) [inline source/common/http/header_map_impl.h]: Dude, Euro are still worth more than dollars! [ ;) ]

(Fixed. Mea culpa, and thanks for the catch!)

File: test/common/http/header_map_impl_test.cc [hunk 2]
```diff
@@ -319,15 +319,126 @@ TEST(HeaderMapImplTest, DoubleInlineAdd) {
   EXPECT_EQ(1UL, headers.size());
 }
 
+struct HeaderBrutalityResult {
```
Comment (mattklein123) [inline test/common/http/header_map_impl_test.cc]: None of this header brutality stuff is really needed to test copying and it makes all of the tests much more confusing. I would recommend removing it all. If you want to make sure copying is working, just allocate string on the heap, add them to headers, delete the string, then access the header map. ASAN will catch any issues.

File: test/common/http/header_map_impl_test.cc [hunk 3]
```diff
@@ -319,15 +319,126 @@ TEST(HeaderMapImplTest, DoubleInlineAdd) {
...
+  // key we're using actually goes out of scope before we check the
+  // HeaderMap for the value we need.
+  LowerCaseString lcKey("hello");
+
+  headers.addCopy(lcKey, 42);
+
+  return doHeaderBrutality(headers, lcKey);
+}
+
+// void dumpHeaders(HeaderMapImpl& headers, int idx) {
```
Comment (mattklein123) [inline test/common/http/header_map_impl_test.cc]: Please delete unused commented out code.

File: source/common/http/header_map_impl.h [hunk 4]
```diff
@@ -54,9 +54,17 @@ class HeaderMapImpl : public HeaderMap {
   bool operator==(const HeaderMapImpl& rhs) const;
 
   // Http::HeaderMap
+  // Add without copying either key or value.
```
Comment (mattklein123) [inline source/common/http/header_map_impl.h]: nit: Please remove the added comments here. In general we don't replicate comments for functions which are already heavily commented in this interface.

File: test/common/http/header_map_impl_test.cc [hunk 5]
```diff
@@ -319,112 +319,54 @@ TEST(HeaderMapImplTest, DoubleInlineAdd) {
...
   HeaderMapImpl headers;
 
   // Start with a string value.
-  HeaderBrutalityResult v1 = addAHeaderString(headers);
-  // dumpHeaders(headers, 1);
-
-  EXPECT_STREQ("world", v1.val.c_str());
-  EXPECT_EQ(5UL, v1.val.size());
-  EXPECT_EQ(1UL, headers.size());
+  LowerCaseString* lcKeyPtr(new LowerCaseString("hello"));
```
Comment (dnoe) [inline test/common/http/header_map_impl_test.cc]: I think it's still better to use unique_ptr here.  It's less important for the test but it's a good habit to use the smart pointer types consistently.

Use `std::unique_ptr<LowerCaseString> lcKeyPtr(new LowerCaseString("hello")` here, and then where you call delete below use `lcKeyPtr.reset()` which will implicitly call delete.

**Comment by dnoe** [comment]
Other than the unique_ptr change and addressing Matt's comment about the implementation method comments I think this LGTM. You'll also need to sign the CLA, visit https://oss.lyft.com/cla for this.

File: test/common/http/header_map_impl_test.cc (hunk 5 above)
Comment (mattklein123) [inline test/common/http/header_map_impl_test.cc]: +1, always use unique_ptr, etc. Thank you. :)

File: test/common/http/header_map_impl_test.cc [hunk 6]
```diff
@@ -323,16 +323,15 @@ TEST(HeaderMapImplTest, AddCopy) {
...
+  std::unique_ptr<LowerCaseString> lcKeyPtr(new LowerCaseString("hello"));
   headers.addCopy(*lcKeyPtr, "world");
 
   const HeaderString& value = headers.get(*lcKeyPtr)->value();
 
   EXPECT_STREQ("world", value.c_str());
   EXPECT_EQ(5UL, value.size());
 
-  delete lcKeyPtr;
+  lcKeyPtr.reset(nullptr);
```
Comment (dnoe) [inline test/common/http/header_map_impl_test.cc]: nit: You can just say `lcKeyPtr.reset()`

File: test/common/http/header_map_impl_test.cc [hunk 7]
```diff
@@ -356,7 +355,7 @@ TEST(HeaderMapImplTest, AddCopy) {
   EXPECT_STREQ("42", value3.c_str());
   EXPECT_EQ(2UL, value3.size());
 
-  delete lcKeyPtr;
+  lcKeyPtr.reset(nullptr);
```
Comment (dnoe) [inline test/common/http/header_map_impl_test.cc]: Same here

**Comment by dnoe** [comment]
You'll still need to do the CLA signing dance before merge, see [CONTRIBUTING.md](https://github.com/lyft/envoy/blob/master/CONTRIBUTING.md) at the bottom.

File: source/common/http/header_map_impl.h [hunk 8]
```diff
@@ -53,10 +53,13 @@ class HeaderMapImpl : public HeaderMap {
    */
   bool operator==(const HeaderMapImpl& rhs) const;
 
-  // Http::HeaderMap
+  // Http::HeaderMap -- read comments there about what's copied and what's not!
```
Comment (mattklein123) [inline source/common/http/header_map_impl.h]: please delete the extra comment here. It's not done anywhere else. The general expectation is that people read the headers.

File: test/common/http/header_map_impl_test.cc [hunk 9]
```diff
@@ -319,15 +319,67 @@ TEST(HeaderMapImplTest, DoubleInlineAdd) {
...
+  const HeaderString& value2 = headers.get(LowerCaseString("hello"))->value();
+
+  EXPECT_STREQ("world", value2.c_str());
+  EXPECT_EQ(5UL, value2.size());
+  EXPECT_EQ(value.c_str(), value2.c_str());
+  EXPECT_EQ(1UL, headers.size());
+
+  // Repeat with an int value.
+  //
+  // @kflynn: addReferenceKey and addCopy can both add multiple instances of a
```
Comment (mattklein123) [inline test/common/http/header_map_impl_test.cc]: Please remove reference to your name, not needed.

File: test/common/http/header_map_impl_test.cc [hunk 10]
```diff
@@ -319,15 +319,67 @@ TEST(HeaderMapImplTest, DoubleInlineAdd) {
...
+  lcKeyPtr.reset(nullptr);
+
+  const HeaderString& value4 = headers.get(LowerCaseString("hello"))->value();
+
+  EXPECT_STREQ("42", value4.c_str());
+  EXPECT_EQ(2UL, value4.size());
+  EXPECT_EQ(1UL, headers.size());
+
+  // We can make yet another different key to look up here.
+  LowerCaseString lcKey3(std::string("he") + "ll" + "o"); // Try to force the compiler's hand here.
```
Comment (mattklein123) [inline test/common/http/header_map_impl_test.cc]: Please remove trailing comment about compiler. Not needed (and not sure exactly what it means).

**Pull request merged**